File: src/KST201Manager.js
import { APTManager } from "./APTManager.js";
import {HW, MOT} from "./util/APT_MGMSG.js"
import { parseStatus } from "./util/APT_STATUS.js";
import { sleep, hdr_short, hdr_long, hdr_w_data, wordLE, longLE } from "./util/util.js";


const USB = 0x50;
const HOST = 0x01;

export class KST201Manager extends APTManager {
  constructor(verbose = 2) {
    super(verbose);


    this.MOVE_HOMED = hdr_short( MOT.MOVE_HOMED, 1, 0, HOST, USB);
    this.MOVE_COMPLETED = hdr_long( MOT.MOVE_COMPLETED, 14, HOST, USB);
    this.GET_STATUSUPDATE = hdr_long( MOT.GET_STATUSUPDATE, 28, HOST, USB);
    // this.MOVE_STOPPED = hdrWithData( MOT.MOVE_STOPPED, 14, HOST, USB);

  }

  async _process_regular(){
    return
  }

  async _process_request(req) {
    const { cmd, data } = req;
    this.log(
      `Process: ${cmd} ${data !== undefined ? JSON.stringify(data) : ""}`,
    );
    await this.purge()
    try {
      if (cmd == "get_status") {
        await this.get_status();
      } else if (cmd === "move_home") {
        await this.move_home();
      } else if (cmd === "move_absolute") {
        await this.move_absolute(data);
      } else if (cmd === "move_relative") {
        await this.move_relative(data);
      } else if (cmd === "move_stop") {
        await this.move_stop();
      } else {
        this.log("Unknown Command: " + cmd);
      }
    } catch (e) {
      console.error("process_request error:", e);
    }
  }

  async get_status() {
    try {
      await this.write_short(MOT.REQ_STATUSUPDATE);
      const resp = await this.waitUntil(this.GET_STATUSUPDATE, 60_000);

      if (resp) {
        const data = await this.readExact(28, 200);
        const pos = data.readInt32LE(2);
        const status = parseStatus(data.readUInt32LE(10));
        status.position = pos;
        console.log(status)
        this.emit("send", status);
      }
      else {
        console.log( hdr)
      }
    } catch (e) {
      console.log("Error: " + e.message);
      return -1;
    }
  }

  async move_stop() {
    try {
      this.log("Stopping");
      await this.write_short(MOT.MOVE_STOP, 1, 1);
    } catch (e) {
      console.log("Error: " + e.message);
      return -1;
    }
  }

  async move_relative(count) {
    try {
      this.log("Move Relative: " + count);
      const data = Buffer.concat([wordLE(1), longLE(count)]); // [Chan WORD][AbsPos LONG]
      await this.write_long(MOT.MOVE_RELATIVE, data);
      const resp = await this.waitUntil(this.MOVE_COMPLETED, 30_000);
      if ( resp ){
        await this.readExact(14,10_000)
        this.get_status()
      }
    } catch (e) {
      console.log("Error: " + e.message);
      return -1;
    }
  }

  async move_absolute(count) {
    try {
      this.log("Move Absolute: " + count);
      const data = Buffer.concat([wordLE(1), longLE(count)]); // [Chan WORD][AbsPos LONG]
      await this.write_long(MOT.MOVE_ABSOLUTE, data);
      const resp = await this.waitUntil(this.MOVE_COMPLETED, 30_000);
      if ( resp ){
        await this.readExact(14)
        this.get_status()
      }
    } catch (e) {
      console.log("Error: " + e.message);
      return -1;
    }
  }

  async move_home() {
    try {
      await this.write_short(MOT.MOVE_HOME);
      const resp = await this.waitUntil(this.MOVE_HOMED, 60_000)
      if ( resp ){
        this.get_status()
      }
    } catch (e) {
      console.log("Error: " + e.message);
      return false;
    }
  }

  async _dev_check() {
    try {
      const serial = await this.get_serial();
      console.log("SERIAL: "+serial)
      if (serial === this.dev_info.SERIAL) {
        this.log("dev_check passed");
        return true;
      } else {
        this.log("dev_check failed");
        return false;
      }
    } catch (e) {
      console.log("Error: " + e.message);
      return false;
    }
  }

  async get_serial() {
    await this.purge();
    await this.write_short(HW.REQ_INFO);

    const hdr = await this.readExact(6);
    if (hdr.readInt16LE(0) === HW.GET_INFO) {
      const data = await this.readExact(84);
      return data.readInt32LE(0);
    } else {
      return -1;
    }
  }

  log(msg) {
    if (this.verbose >= 1) {
      console.log(`[KST201Manager]: ${msg}`);
    }
  }
}

File: src/KST201Manager.py
from APTManager import APTManager
from time import sleep
from util import hdr_long, hdr_short, hdr_with_data
from datetime import datetime
from status import parseStatus

HW_REQ_INFO = 0x0005
HW_GET_INFO = 0x0006

MOT_MOVE_HOME = 0x0443
MOT_MOVE_HOMED = 0x0444
MOT_MOVE_RELATIVE = 0x0448
MOT_SET_MOVEABSPARAMS = 0x0450
MOT_REQ_MOVEABSPARAMS = 0x0451
MOT_GET_MOVEABSPARAMS = 0x0452
MOT_MOVE_ABSOLUTE = 0x0453
MOT_MOVE_COMPLETED = 0x0464
MOT_MOVE_STOP = 0x0465
MOT_MOVE_STOPPED = 0x0466
MOT_REQ_STATUSUPDATE = 0x0480
MOT_GET_STATUSUPDATE = 0x0481

USB = 0x50
HOST = 0x01

class KST201Manager(APTManager):
    def __init__( self, verbose=2 ):
        super().__init__(verbose)

        self.GET_INFO = hdr_long( HW_GET_INFO, 84, HOST, USB)
        self.GET_STATUSUPDATE = hdr_long( MOT_GET_STATUSUPDATE, 28, HOST, USB)
        self.MOVE_HOMED = hdr_short( MOT_MOVE_HOMED,1,0,HOST,USB)
        self.MOVE_COMPLETED = hdr_long( MOT_MOVE_COMPLETED,14,HOST,USB)


    def _dev_check(self):
        serial = self.get_serial()

        if (serial == self.dev_info['SERIAL']) :
            print( "dev_check passed")
            return True
        else :
            print( "dev_check failed")
            return False
        
    def get_serial(self):
        self.purge()
        self.write_short( HW_REQ_INFO )
        sleep( 0.3 )
        resp = self.read( 6 )
        if (resp== self.GET_INFO) :
            data = self.read(84)
            serial = int.from_bytes(data[0:4], 'little', signed=False)
            print( f"SERIAL: {serial}" )
        else :
            serial = -1
            print( f"INVALID SERIAL: {serial}" )
        return serial


    def get_status(self):
        self.purge()
        self.write_short( MOT_REQ_STATUSUPDATE )
        sleep(0.3)

        stat = self.wait_until( self.GET_STATUSUPDATE, 60)
        if (stat):
            data = self.read(28)
            pos = int.from_bytes(data[2:10],'little', signed=True)
            stat = int.from_bytes(data[10:28],'little', signed=False)

            status = parseStatus(stat)
            status['position'] = pos
            print(f"STATUS: {status}")
            return status


    def move_home(self):
        self.purge()
        self.write_short( MOT_MOVE_HOME)

        homed = self.wait_until( self.MOVE_HOMED, 60)
        print( f"HOMED: {homed}")
        status = self.get_status()
        return status
    
    def move_absolute(self,target):
        self.purge()
        chan = (1).to_bytes(2,'little')
        pos = int(target).to_bytes(4,'little',signed=True)
        payload = chan+pos

        self.write_long( MOT_MOVE_ABSOLUTE,payload )
        moved = self.wait_until( self.MOVE_COMPLETED, 60)
        print( f"MOVED: {moved}")
        if moved:
            data = self.read(14)
            pos = int.from_bytes(data[2:6],'little', signed=True)
            stat = int.from_bytes(data[10:14],'little', signed=False)

            status = parseStatus(stat)
            status['position'] = pos
            print(f"STATUS: {status}")
            return status
        return moved
    
    def move_relative(self,count):
        self.purge()
        chan = (1).to_bytes(2,'little')
        pos = int(count).to_bytes(4,'little',signed=True)
        payload = chan+pos

        self.write_long( MOT_MOVE_RELATIVE,payload )
        moved = self.wait_until( self.MOVE_COMPLETED, 60)
        print( f"MOVED: {moved}")
        if moved:
            data = self.read(14)
            pos = int.from_bytes(data[2:6],'little', signed=True)
            stat = int.from_bytes(data[10:14],'little', signed=False)

            status = parseStatus(stat)
            status['position'] = pos
            print(f"STATUS: {status}")
            return status
        return moved

    def wait_until(self, header, timeout=3):
        deadline = datetime.now().timestamp()+timeout
        while ( deadline > datetime.now().timestamp()):
            resp = self.read(len(header))
            if (resp == header):
                return True
            elif (len(resp)>0):
                print( f"Wrong Header: {resp}")
            
        print("Timed out during wait_until")
        return False
        
